test/Npgsql.PluginTests/JsonNetTests.cs

```diff
@@ -54,6 +56,40 @@ public void DeserializeFailure()
             }
         }
 
+        [Test, IssueLink("https://github.com/npgsql/npgsql/issues/3085")]
+        public void RoundtripStringTypes()
+        {
+            var expected1 = @"{""Key"": ""Value""}";
+            var expected2 = new char[] { '{', '"', 'p', '"',':', '1', '}' };
+            var expected3 = new ArraySegment<char>("{\"p\":1}".ToCharArray());
+            // If we serialize to JSONB, Postgres will not store the Json.NET formatting, and will add a space after ':'
+            var expected2String = _npgsqlDbType.Equals(NpgsqlDbType.Jsonb) ? "{\"p\": 1}"
+                                    : "{\"p\":1}";
+            var expected4 = Encoding.ASCII.GetBytes("\"test\"");
+
+            using var conn = OpenConnection();
+            using var cmd = new NpgsqlCommand(@"SELECT @p1, @p2, @p3, @p4", conn);
+
+            cmd.Parameters.Add(new NpgsqlParameter<string>("p1", _npgsqlDbType) { Value = expected1 });
+            cmd.Parameters.Add(new NpgsqlParameter<char[]>("p2", _npgsqlDbType) { Value = expected2 });
+            cmd.Parameters.Add(new NpgsqlParameter<ArraySegment<char>>("p3", _npgsqlDbType) { Value = expected3 });
+            cmd.Parameters.Add(new NpgsqlParameter<byte[]>("p4", _npgsqlDbType) { Value = expected4 });
+
+            try
+            {
+                using var reader = cmd.ExecuteReader();
+                reader.Read();
+                Assert.That(reader.GetFieldValue<string>(0), Is.EqualTo(expected1));
+                Assert.That(reader.GetFieldValue<string>(1), Is.EqualTo(expected2String));
+                Assert.That(reader.GetFieldValue<string>(2), Is.EqualTo(expected2String));
+                Assert.That(reader.GetFieldValue<string>(3), Is.EqualTo(Encoding.ASCII.GetString(expected4)));
```

Comment:
While this test verifies that the handlers write values correctly, it doesn't check reading.

Reply:
Ok, thanks for clarifying that. I had a misunderstanding on how the GetFieldValue<> function worked and thought there was an issue with my test so changed them to get the string type. Thanks for pointing this out as what I thought I was doing incorrectly is actually a bug in Npgsql.Json.Net.JsonHandler's Read<T> method which was throwing the following exception when reading for char[] type:

```
Newtonsoft.Json.JsonSerializationException: 'Cannot deserialize the current JSON object (e.g. {"name":"value"}) into
type 'System.Char[]' because the type requires a JSON array (e.g. [1,2,3]) to deserialize correctly.
To fix this error either change the JSON to a JSON array (e.g. [1,2,3]) or change the deserialized type so that it is a 
normal .NET type (e.g. not a primitive type like integer, not a collection type like an array or List<T>) that can be 
deserialized from a JSON object.
JsonObjectAttribute can also be added to the type to force it to deserialize from a JSON object.
```
I'm currently looking into this and I think I have a fix (based on the same function which works on the JsonbHandler). 

Also, I just wanted to confirm that reading from type `ArraySegment<char>` is currently not supported as I get the following exception thrown:

`System.NotSupportedException: 'Only writing ArraySegment<char> to PostgreSQL text is supported, no reading.'`

Checking the code, the exception is being explicitly thrown by the TextHandler's `INpgsqlTypeHandler<ArraySegment<char>>.Read` method. 

For the moment I'll pull out the Read test for the `ArraySegment<char>` type and implement it in it's own test with an ignore (not implemented yet) attribute on the test?

Reply:
Sure, as you wish. The same test should be added to the standard handler too, just copy paste the code to the main test project.

Reply:
Good point, I'll do that now.

Quick question around the test, if I push them now, the CI tests will fail because of the bug I found reading the char[] type on the jsonhandler. Should I push the fix for this before I push the test changes? or push the test updates (which will cause CI tests to fail) first?

Reply:
Since build should not fail on main branches, it would be better to add those parts which will pass now and fix with a test later. Otherwise, you can push a fix testing only that part and rebase this PR.

The first is preferable I think since it requires less actions for you (: